python/graphstorm/model/node_glem.py

```diff
@@ -0,0 +1,336 @@
+        self.lm.set_loss_func(loss_fn)
+        self.gnn.set_loss_func(loss_fn)
+
+    def freeze_params(self, part='lm'):
+        """Freeze parameters in lm or gnn"""
+        if part == 'lm':
+            params = self.lm.parameters()
+        elif part == 'gnn':
+            params = self.gnn.parameters()
+        for param in params:
+            param.requires_grad = False
+
+    def unfreeze_params(self, part='lm'):
+        """Unfreeze parameters in lm or gnn"""
+        if part == 'lm':
+            params = self.lm.parameters()
+        elif part == 'lm-input-proj':
+            params = self.lm.node_input_encoder.input_projs.parameters()
+        elif part == 'gnn':
+            params = self.gnn.parameters()
+        for param in params:
+            param.requires_grad = True
+
+    def toggle(self, part='lm'):
+        """The method toggles training between lm and gnn."""
+        if part == 'lm':
+            self.training_lm = True
+            self.freeze_params('gnn')
+            self.unfreeze_params('lm')
+        elif part == 'gnn':
+            self.training_lm = False
+            self.freeze_params('lm')
+            self.unfreeze_params('gnn')
+            self.unfreeze_params('lm-input-proj')
+        else:
+            raise ValueError(f"Unknown model part: {part}")
+
+    def forward(self, blocks, node_feats, edge_feats, labels, input_nodes, use_gnn=True,
+                no_pl=False):
+        """ Forward pass for GLEM model.
+        Parameters
+        ----------        
+        blocks : List[dgl.heterograph.DGLBlock]
+        node_feats : Dict[ntype: tensors.shape [bs, feat_dim]]
+        edge_feats : None
+        labels : Dict[target_ntype: tensor.shape [bs]]
+        input_nodes : {target_ntype: tensor.shape [bs], other_ntype: []}
+        use_gnn : bool
+            If True, use GNN's decoder, otherwise, use LM's decoder
+        no_pl : bool
+            If True, do not calculate pseudo likelihood, use MLE loss only
+        """
+        if use_gnn:
+            total_loss = self.forward_gnn(blocks, node_feats, edge_feats, labels, input_nodes,
+                                          no_pl=no_pl)
+        else:
+            total_loss = self.forward_lm(blocks, node_feats, edge_feats, labels, input_nodes,
+                                         no_pl=no_pl)
+        return total_loss
+
+    def forward_gnn(self, blocks, node_feats, _, labels, input_nodes, no_pl=False):
+        """Forward pass for node prediction using GNN
+        """
+        emb_lm, emb_gnn = self._embed_nodes(blocks, node_feats, _, input_nodes, do_gnn_encode=True)
+        labels = self._process_labels(labels)
+        logits = self.gnn.decoder(emb_gnn)
+        if no_pl:
+            loss = self.gnn.loss_func(logits, labels)
+        else:
+            batch_size = labels.size(0)
+            n_gold_nodes = batch_size // 2
+            # compute pseudo labels from LM
+            logits_lm = self.lm.decoder(emb_lm[n_gold_nodes:])
```

Comment:
this PR uses the existing training set to train the GNN model with LM or vice versa. do we expect to get better performance if we take advantage of the unlabelled nodes? does the original paper only use the training set to train the model?

Reply:
We expect to get better performance if we use unlabeled nodes. The original paper uses unlabeled nodes in transductive setting. They also have experiments in inductive setting.